src/ir/term/mod.rs

```diff
@@ -1252,175 +1258,227 @@ impl Value {
-                    BvBinPred::Slt => a.as_sint() < b.as_sint(),
-                    BvBinPred::Uge => a.uint() >= b.uint(),
-                    BvBinPred::Ugt => a.uint() > b.uint(),
-                    BvBinPred::Ule => a.uint() <= b.uint(),
-                    BvBinPred::Ult => a.uint() < b.uint(),
-                }
-            }),
-            Op::BoolToBv => Value::BitVector(BitVector::new(
-                Integer::from(vs.get(&c.cs[0]).unwrap().as_bool()),
-                1,
-            )),
-            Op::PfUnOp(o) => Value::Field({
-                let a = vs.get(&c.cs[0]).unwrap().as_pf().clone();
+        }),
+        Op::BoolToBv => Value::BitVector(BitVector::new(
+            Integer::from(vs.get(&c.cs[0]).unwrap().as_bool()),
+            1,
+        )),
+        Op::PfUnOp(o) => Value::Field({
+            let a = vs.get(&c.cs[0]).unwrap().as_pf().clone();
+            match o {
+                PfUnOp::Recip => a.recip(),
+                PfUnOp::Neg => -a,
+            }
+        }),
+        Op::PfNaryOp(o) => Value::Field({
+            let mut xs = c.cs.iter().map(|c| vs.get(c).unwrap().as_pf().clone());
+            let f = xs.next().unwrap();
+            xs.fold(
+                f,
                 match o {
-                    PfUnOp::Recip => a.recip(),
-                    PfUnOp::Neg => -a,
+                    PfNaryOp::Add => std::ops::Add::add,
+                    PfNaryOp::Mul => std::ops::Mul::mul,
+                },
+            )
+        }),
+        Op::UbvToPf(m) => Value::Field({
+            let a = vs.get(&c.cs[0]).unwrap().as_bv().clone();
+            field::FieldElem::new(a.uint().clone(), m.clone())
+        }),
+        // tuple
+        Op::Tuple => Value::Tuple(c.cs.iter().map(|c| vs.get(c).unwrap().clone()).collect()),
+        Op::Field(i) => {
+            let t = vs.get(&c.cs[0]).unwrap().as_tuple();
+            assert!(i < &t.len(), "{} out of bounds for {}", i, c.cs[0]);
+            t[*i].clone()
+        }
+        Op::Update(i) => {
+            let mut t = Vec::from(vs.get(&c.cs[0]).unwrap().as_tuple()).into_boxed_slice();
+            assert!(i < &t.len(), "{} out of bounds for {}", i, c.cs[0]);
+            let e = vs.get(&c.cs[1]).unwrap().clone();
+            assert_eq!(t[*i].sort(), e.sort());
+            t[*i] = e;
+            Value::Tuple(t)
+        }
+        // array
+        Op::Store => {
+            let a = vs.get(&c.cs[0]).unwrap().as_array().clone();
+            let i = vs.get(&c.cs[1]).unwrap().clone();
+            let v = vs.get(&c.cs[2]).unwrap().clone();
+            Value::Array(a.clone().store(i, v))
+        }
+        Op::Select => {
+            let a = vs.get(&c.cs[0]).unwrap().as_array().clone();
+            let i = vs.get(&c.cs[1]).unwrap();
+            a.clone().select(i)
+        }
+        Op::Map(op) => {
+            let arg_cnt = c.cs.len();
+            let arr_size = vs.get(&c.cs[0]).unwrap().as_array().size;
```

Comment:
I think that you should explicitly check the key sort so that you can generate a good error message. See comment below for permitted key sorts.

Reply:
Won't this check be done by the type checking functions?

Reply:
Good point!